crates/miden-lib/src/note/well_known_note.rs

```diff
@@ -157,4 +161,145 @@ impl WellKnownNote {
+                    // - target account ID could be equal to the inputs account ID if the note is
+                    //   going to be consumed by the target account
+                    // - target account ID could be equal to the sender account ID if the note is
+                    //   going to be consumed by the sender account
+                    if [input_account_id, sender_account_id].contains(&target_account_id) {
+                        None
+                    } else {
+                        Some(NoteConsumptionStatus::Incompatible("Target account of the transaction does not match neither the target account specified by the P2IDE inputs, nor the sender account".to_string()))
+                    }
+                } else {
+                    // in this case note could be consumed only by the target account
+                    if input_account_id == target_account_id {
+                        None
+                    } else {
+                        Some(NoteConsumptionStatus::ConsumableAfter(BlockNumber::from(
+                            recall_height,
+                        )))
+                    }
+                }
+            },
+            _ => None,
+        }
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// Reads the account ID from the first two note input values.
+///
+/// Returns None if the note input values used to construct the account ID are invalid.
+fn try_read_account_id_from_inputs(note_inputs: &[Felt]) -> Option<AccountId> {
+    let account_id_felts: [Felt; 2] = note_inputs[0..2].try_into().expect(
+        "Should be able to convert the first two note inputs to an array of two Felt elements",
+    );
+
+    AccountId::try_from([account_id_felts[1], account_id_felts[0]]).ok()
+}
+
+// HELPER STRUCTURES
+// ================================================================================================
+
+/// Describes if a note could be consumed under a specific conditions: target account state
+/// and block height.
+///
+/// The status does not account for any authorization that may be required to consume the
+/// note, nor does it indicate whether the account has sufficient fees to consume it.
+#[derive(Debug, PartialEq)]
+pub enum NoteConsumptionStatus {
+    /// The note can be consumed by the account at the specified block height.
+    Consumable,
+    /// The note can be consumed by the account after the required block height is achieved.
+    ConsumableAfter(BlockNumber),
+    /// The note can be consumed by the account if proper authorization is provided.
+    ConsumableWithAuthorization,
+    /// The note cannot be consumed by the account at the specified conditions (i.e., block
+    /// height and account state).
+    Unconsumable,
+    /// The note cannot be consumed by the specified account under any conditions.
+    Incompatible(String),
```

Comment:
I think I would change this to `Box<dyn Error + Send + Sync + 'static>`, as that seems more flexible for future scenario where we may not want to stringify a potential `Error` that we have.